fastlane_core/lib/fastlane_core/analytics/analytics_session.rb

```diff
@@ -0,0 +1,197 @@
+        session_id: session_id,
+        action_name: launch_context.action_name
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'fastlane_version',
+          detail: fastlane_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'install_method',
+          detail: install_method
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'operating_system',
+          detail: operating_system
+        },
+        secondary_target_hash: {
+          name: 'version',
+          detail: operating_system_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'build_tool_version',
+          detail: launch_context.build_tool_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'ci',
+          detail: ci?.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'fastfile',
+          detail: fastfile?.to_s
+        },
+        secondary_target_hash: {
+          name: 'fastfile_id',
+          detail: fastfile_id.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'platform',
+          detail: launch_context.platform.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'ruby_version',
+          detail: ruby_version
+        }
+      )
+    end
+
+    def is_fastfile=(value)
+      if value
+        # If true, update all of the events to reflect
+        # that the execution is running within a Fastfile context.
+        # We don't want to update if this is false because once we
+        # detect a true value, that is the one to be trusted
+        @events.reverse_each do |event|
+          event[:primary_target][:name] == 'fastfile' ? event[:primary_target][:detail] = value.to_s : next
+        end
+      end
+      @is_fastfile = value
+    end
+
+    def action_completed(completion_context: nil)
+      backfill_p_hashes(p_hash: completion_context.p_hash)
+
+      builder = AnalyticsEventBuilder.new(
+        oauth_app_name: oauth_app_name,
+        p_hash: completion_context.p_hash,
+        session_id: session_id,
+        action_name: completion_context.action_name
+      )
+
+      @events << builder.completed_event(
+        primary_target_hash: {
+          name: 'status',
+          detail: completion_context.status
+        }
+      )
+    end
+
+    def finalize_session
+      # If our users want to opt out of usage metrics, don't post the events.
+      # Learn more at https://github.com/fastlane/fastlane#metrics
+      return if FastlaneCore::Env.truthy?("FASTLANE_OPT_OUT_USAGE")
+
+      client.post_events(@events)
```

Comment:
ONE PLACE TO POST EVENTS!!!! Yessssssss the best. So much easier to determine where we send stuff